### PR TITLE
Lint: filter(Option::is_some).map(Option::unwrap)

### LINKED ISSUE #6061: Suggest replacing filter-for-some-and-unwrap with .flatten
### What it does

Suggests replacing code that filters with `Option::is_some` and then maps `Option::unwrap` with `.flatten()` which does the same thing due to `Option` implementing `IntoIterator`.

### Categories (optional)

- Kind: clippy::style

### Advantage

It is one function call fewer and makes it clearer what is happening - one level of indirection is being removed. In addition, many codebases want to eliminate all possible panic sources, of which Option::unwrap is one, although in this case it is statically known not to panic.

### Drawbacks

None.

### Example

```rust
fn odds_out(x: i32) -> Option<i32> { if x % 2 == 0 { Some(x) } else { None } }
let evens: Vec<_> = myints.iter().map(odds_out).filter(Option::is_some).map(Option::unwrap).collect();
```

Could be written as:

```rust
fn odds_out(x: i32) -> Option<i32> { if x % 2 == 0 { Some(x) } else { None } }
let evens: Vec<_> = myints.iter().map(odds_out).flatten().collect();
```

### PR DESCRIPTION
**Pull request by bbqbaron**
Fixes #6061

*Please write a short comment explaining your change (or "none" for internal only changes)*
changelog: 
* add new lint for filter(Option::is_some).map(Option::unwrap)

First Rust PR, so I'm sure I've violated some idioms. Happy to change anything.

I'm getting one test failure locally -- a stderr diff for `compile_test`. I'm having a hard time seeing how I could be causing it, so I'm tentatively opening this in the hopes that it's an artifact of my local setup against `rustc`. Hoping it can at least still be reviewed in the meantime.

I'm gathering that since this is a method lint, and `.filter(...).map(...)` is already checked, the means of implementation needs to be a little different, so I didn't exactly follow the setup boilerplate. My way of checking for method calls seems a little too direct (ie, "is the second element of the expression literally the path for `Option::is_some`?"), but it seems like that's how some other lints work, so I went with it. I'm assuming we're not concerned about, eg, closures that just end up equivalent to `Option::is_some` by eta reduction.

### CODE REVIEW

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -816,6 +817,28 @@ declare_clippy_lint! {
...
+    ///
+    /// ```rust
+    /// let _ = std::iter::empty::<Option<i32>>().filter(Option::is_some).map(Option::unwrap);
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let _ = std::iter::empty::<Option<i32>>().flatten();
+    /// ```
+    pub OPTION_FILTER_MAP,
+    style,
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:838]:
```suggestion
    complexity,
```

You have to rerun `cargo dev update_lints` after this change.

File: clippy_lints/src/methods/option_filter_map.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+use rustc_hir::{self, Expr, ExprKind};
+use std::iter;
+
+fn calls_option_method<'tcx>(expr: &'tcx [Expr<'_>], fn_name: &str) -> bool {
+    expr.get(1).map_or(false, |x| match &x.kind {
+        ExprKind::Path(qp) => match_qpath(
+            qp,
+            &paths::OPTION
+                .iter()
+                .cloned()
```
Comment (flip1995) [inline clippy_lints/src/methods/option_filter_map.rs:11]: Isn't

```suggestion
                .copied()
```

enough here?

File: clippy_lints/src/methods/option_filter_map.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+                .chain(iter::once(fn_name))
+                .collect::<Vec<&str>>(),
+        ),
+        _ => false,
+    })
+}
+
+pub fn in_scope<'tcx>(filter_args: &'tcx [Expr<'_>], map_args: &'tcx [Expr<'_>]) -> bool {
```
Comment (flip1995) [inline clippy_lints/src/methods/option_filter_map.rs:19]: This doesn't have to be in its own module. And NIT: the `in_scope` function name doesn't really make sense to me. Maybe move it to `mod.rs` and call it `is_option_filter_map`.

File: clippy_lints/src/methods/mod.rs [hunk 4]
```diff
@@ -2895,14 +2919,26 @@ fn lint_skip_while_next<'tcx>(
...
+            cx,
+            OPTION_FILTER_MAP,
+            expr.span,
+            "`filter` for `Some` followed by `unwrap`",
+            None,
+            "consider using `flatten` instead",
+        );
+    } else {
+        // lint if caller of `.filter().map()` is an Iterator
+        if match_trait_method(cx, expr, &paths::ITERATOR) {
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:2937]: This check is also important for the new lint.

File: clippy_lints/src/methods/mod.rs [hunk 5]
```diff
@@ -2895,14 +2919,26 @@ fn lint_skip_while_next<'tcx>(
...
+    // is it specifically `.filter(Option::is_some).map(Option::unwrap)`?
+    if option_filter_map::in_scope(filter_args, map_args) {
+        span_lint_and_help(
+            cx,
+            OPTION_FILTER_MAP,
+            expr.span,
+            "`filter` for `Some` followed by `unwrap`",
+            None,
+            "consider using `flatten` instead",
+        );
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:2934]: This should be `span_lint_and_sugg`. You have to modify the span for this: `filter_span.with_hi(expr.span.hi())`. You can pass the `filter_span` to this function from the colossal match from above.

**Review by flip1995 — COMMENTED** [review]
LGTM. Some code cleanup left to do. 

To also have this pass locally: rebase on `master` and rerun `setup-toolchain.sh`

**Comment by camsteffen** [comment]
Consider supporting/testing these cases:

```rust
// Option<Option<_>>
let _ = Some(Some(1)).filter(Option::is_some).map(Option::unwrap);
// lambdas
let _ = vec![Some(1)].into_iter().filter(|o| o.is_some()).map(|o| o.unwrap());
```

**Comment by bbqbaron** [comment]
Still alive and will do these extra cases/cleanups. Big work thing came up, but soon!

**Comment by bbqbaron** [comment]
Made several of the required changes, but conflict resolution against master seems to have gone sideways. Cleaning up.

**Comment by bbqbaron** [comment]
Right, so, assuming the diff view fixes itself:

* now handling closures that call the appropriate method. I cribbed from code with a [similar need](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/methods/unnecessary_filter_map.rs#L55); I hope I inferred the methodology right.
* handle options as well via `is_type_diagnostic_item`
* switch to suggestions and test fixes
* add some negative tests, as it were, that `.map(blah).flatten()` comes out of this stage (presumably to be linted by another stage)

The extra scope meant I had to guess at best practices for a few things; I hope picked the right methods. I'm used to techniques from simpler languages where you can just beta-reduce and compare normal forms, which I understand can't work easily in Rust.

File: clippy_lints/src/methods/mod.rs [hunk 6]
```diff
@@ -2908,18 +2932,85 @@ fn lint_skip_while_next<'tcx>(
...
+                    .map_or(false, |expr| is_method(cx, &expr, method_path, method_name)),
+                _ => false,
+            }
+        },
+        _ => false,
+    }
+}
+
+// these two paths can't be checked and thus aren't in `paths`
+const OPTION_IS_SOME: [&str; 4] = ["core", "option", "Option", "is_some"];
```
Comment (bbqbaron) [inline clippy_lints/src/methods/mod.rs:2965]: I added these just in time to collide with the path-checking dogfooding we added. They're AFAIK not checkable as paths, so I put them here. Thoughts? Top of this file?

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:2965]: Why aren't those not checkable by the internal lint? Either this is a bug in the internal lint or the paths are just wrong.

File: clippy_lints/src/methods/mod.rs [hunk 7]
```diff
@@ -2908,18 +2932,85 @@ fn lint_skip_while_next<'tcx>(
     }
 }
 
-/// lint use of `filter().map()` for `Iterators`
+fn is_method<'tcx>(cx: &LateContext<'tcx>, expr: &hir::Expr<'_>, method_path: &[&str], method_name: Symbol) -> bool {
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:2935]: Something similar is also done in `bind_instead_of_map` and other modules. In the `Closure` case you should call 

```rust
let closure_body = cx.tcx.hir().body(body_id);
let closure_expr = remove_blocks(&closure_body.value);
```

and then just match the `closure_expr` against `ExprKind::MethodCall`. On `args[0]` you can just check with `is_type_diagnostic_item`.

**Review by flip1995 — CHANGES_REQUESTED** [review]
I wouldn't add a new lint for this, but just reuse the `FILTER_MAP` lint.

**Comment by bbqbaron** [comment]
I'm noticing #6453 overlaps _heavily_ with this, due to #3188 containing elements of #6061. I'm not sure which of the reviewers/submitters have more background here, or if I should just sort it out with #6453's author? @flip1995 what do you think? I'll also tag @camsteffen as the author of #6453.

(I have no attachment to being the one to fix it, and I don't want to fall victim to any sunk-cost fallacies. If I should just close this narrower solution, that's fine.)

**Comment by camsteffen** [comment]
@bbqbaron Sorry I didn't anticipate that since my last comment. I think we haven't done any overlapping work *yet*. But I agree, my PR will supercede most of this, as it is currently planned. I think you could either A) just lint the non-closure case since that is the one non-overlapping piece or B) extend this PR to do what is currently planned in #6453. I have no problem with that since I haven't started that work yet and you opened this PR first. With option B, #6453 can just become `filter_map_more` as discussed there.

**Comment by flip1995** [comment]
@camsteffen @bbqbaron It doesn't seem that there is any overlapping work yet. But #6453 will change the lint significantly, especially once mikerites suggestion is implemented (which I think we want to do?).

So I'd encourage you to work together to improve this lint in general or to at least wait for the changes in #6453 before continue working on this. That will avoid rebase/merge conflicts between the two PRs.

I think the changes made here can be used as-is in #6453 to implement the first case in mikerites suggestion.

**Comment by bbqbaron** [comment]
Ok, so if in doubt I'll at least hold off for #6453, or once the holidays are over, if for some reason we need to intermingle the changes, I'm around to collaborate.

**Comment by giraffate** [comment]
FYI https://github.com/rust-lang/rust-clippy/pull/6453 is closed, but the replacing PR https://github.com/rust-lang/rust-clippy/pull/6591 is opened and already merged.

**Comment by camsteffen** [comment]
Thanks @giraffate! I forgot to follow up.

@bbqbaron You may pick this back up if you'd like. My PR did not cover the non-closure case. FYI, I have a shared implementation for `manual_filter_map` and `manual_find_map` which you can extend on to improve both lints. It would be nice to cover any combination of closure and non-closure args to filter and and map.

File: clippy_lints/src/methods/mod.rs (hunk 6 above)
Comment (camsteffen) [reply clippy_lints/src/methods/mod.rs:2965]: Ah! I fixed this in #6567.

**Comment by camsteffen** [comment]
Thought about this a little more. This is still a new lint and doesn't overlap a whole lot with the implementation of my PR. So just want to remove the expectation that this should "extend" on #6591.

I have another thought. Would it make more sense to simply lint `filter(Option::is_some)` and call the lint `filter_some`? To me that seems lint-able without looking at what comes after `filter` (or `find`).

**Comment by bbqbaron** [comment]
@camsteffen thanks for the ping and sorry for the delay. as a first-timer on the repo i'm a little unsure who's ultimately responsible for decisions, but 

> I have another thought. Would it make more sense to simply lint filter(Option::is_some) and call the lint filter_some? To me that seems lint-able without looking at what comes after filter (or find).

...makes some sense to me. i suppose there are legitimate use cases for partitioning optionals into present and absent cases, though i can't come up with an immediately obvious one. 

hm. to me, the more suspicious expression is the `map(Option::unwrap)`, which seems to me more likely to be a bad idea that could be replaced by another methodology (although i'm not a Rust veteran). what do you think?

**Comment by camsteffen** [comment]
> @camsteffen thanks for the ping and sorry for the delay.

No worries!

> i suppose there are legitimate use cases for partitioning optionals into present and absent cases...

Maybe this is just semantics, but filtering is different from partitioning. With `filter(Option::is_some)`, the user is just throwing away `None`s without observing them in any way. So there is no information lost by changing to `flatten()`.

The only somewhat questionable case I can think of is `filter(Option::is_some).for_each(fn_takes_option)`. But even then I think `flatten().for_each(|x| fn_takes_option(Some(x))` is better for clarity.

> the more suspicious expression is the `map(Option::unwrap)`

I disagree. It could be bad error handling (just like any usage of `unwrap`), but there may be a perfectly valid reason to expect values to be `Some` (again, just like any usage of `unwrap`). Example:

```rust
// I can safely unwrap here
"bad".chars().map(|c| "abcd".find(c)).map(Option::unwrap)
```

With that said, I don't want to put up a road block if you would like to go ahead with the original intent of this PR.

**Comment by giraffate** [comment]
ping from triage @bbqbaron. Do you need help to move this forward? If so, feel free to ask reviewers questions.

**Comment by bbqbaron** [comment]
> ping from triage @bbqbaron. Do you need help to move this forward? If so, feel free to ask reviewers questions.

Thanks for pinging me. Last time I picked this up, I ran into some build errors, probably due to movement in the toolchain underneath me. I took too long before coming back. I've got it building now and will resume (hopefully not restart!) implementation, targeting an updated PR this weekend.

**Comment by bbqbaron** [comment]
i've updated to incorporate all the changes to master that this branch has missed. i think there are a couple of items to confirm:

* discussion on #6591 suggests that `option_filter_map` _should_ be a new lint, counter to [suggestion](https://github.com/rust-lang/rust-clippy/pull/6342#pullrequestreview-555507384) above, so i left it in, as `filter_map` seems to be getting decomposed.

* so i could get changes up and confirm scope/intended behavior, i've just shoehorned `option_filter_map` into the middle of the `lint_filter_map` function introduced in #6591. tests pass, although it's kind of crude. they could be integrated more cleanly. however, given all the back-and-forth on #6591 and its ancestors/tickets, i wanted to make sure i hadn't missed/misinterpreted important movement/discussion:

it sounds like, to repeat it back and make sure i have it right, the only real new functionality now covered by #6061 is that a `filter->map` that is exactly `is_some` into `unwrap` (either via direct method reference or trivial closure) should become `flatten`, which is a little cleaner than an identity `flat_map`. we already have a lint that tries to reject trivial closures (`|x| some_fn(x)`), so depending on whether we assume that lint is active, we could literally just check for method references (`Option::is_some`) and nothing else, delegating anything more complicated to the excellent work on #6591?

sorry, this ticket feels like it's cost a lot more discussion/effort than the ultimately quite limited change it makes, if i have all this right. apologies! new to non-toy Rust and the repo, so hopefully any future changes will be smoother.

**Comment by giraffate** [comment]
Can you do a rebase, not merge? We follow a no-merge policy: <https://github.com/rust-lang/rust-clippy/blob/master/doc/basics.md#pr>.

**Comment by bbqbaron** [comment]
> Can you do a rebase, not merge? We follow a no-merge policy: https://github.com/rust-lang/rust-clippy/blob/master/doc/basics.md#pr.

oh, sorry, even when merging against my own branch? sure. i think i can...back out the merge commit and force-push a rebase. will give it a shot.

File: tests/ui/option_filter_map.stderr [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+LL |     let _ = Some(Some(1)).filter(Option::is_some).map(Option::unwrap);
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using `flatten` instead: `Some(Some(1)).flatten()`
+   |
+   = note: `-D clippy::option-filter-map` implied by `-D warnings`
+
+error: `filter` for `Some` followed by `unwrap`
+  --> $DIR/option_filter_map.rs:9:13
+   |
+LL |     let _ = Some(Some(1)).filter(|o| o.is_some()).map(|o| o.unwrap());
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using `flatten` instead: `Some(Some(1)).flatten()`
```
Comment (flip1995) [inline tests/ui/option_filter_map.stderr:13]: For a nicer suggestion with longer method chains, you shouldn't include the iterator filter and map are called on, but just the span of `filter().map()`

File: clippy_lints/src/methods/mod.rs (hunk 7 above)
Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:3191]: I think my [previous comment](https://github.com/rust-lang/rust-clippy/pull/6342/files#r545862641) still applies here. This closure unpacking now appears in 4 lints across 4 files. This should probably be moved to the methods/utils module. But this should be done in another PR.

File: clippy_lints/src/methods/mod.rs [hunk 9]
```diff
@@ -3089,10 +3114,11 @@ fn lint_skip_while_next<'tcx>(
 }
 
 /// lint use of `filter().map()` or `find().map()` for `Iterators`
-fn lint_filter_map<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>, is_find: bool) {
+fn lint_filter_map<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>, is_find: bool, target_span: Span) {
     if_chain! {
         if let ExprKind::MethodCall(_, _, [map_recv, map_arg], map_span) = expr.kind;
-        if let ExprKind::MethodCall(_, _, [_, filter_arg], filter_span) = map_recv.kind;
+        if let ExprKind::MethodCall(_, _, [filter_recv, filter_arg], filter_span) = map_recv.kind;
+        let _ = lint_filter_some_map_unwrap(cx, expr, filter_recv, filter_arg, map_arg, target_span);
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:3121]: Sneaking this into the `if_chain` seems very wrong to me.

File: clippy_lints/src/methods/bind_instead_of_map.rs [hunk 10]
```diff
@@ -1,7 +1,8 @@
-use super::{contains_return, BIND_INSTEAD_OF_MAP};
+use super::BIND_INSTEAD_OF_MAP;
```
Comment (flip1995) [inline clippy_lints/src/methods/bind_instead_of_map.rs:1]: Those changes seem unrelated

**Review by flip1995 — CHANGES_REQUESTED** [review]
Sorry, I totally forgot about this PR.

The methods module is now split up even further and each lint is now in its own module. Also in the last 1-2 weeks much refactoring was done in Clippy, so you probably will have to rebase this PR and fix some imports etc.

**Comment by bbqbaron** [comment]
Thanks for the notes. I got moderately severe covid, so I'll be gone
for...well, who knows.

I'm worried there'll be yet another new round of major conflicts when I
return, but I guess we'll see!

**Comment by bbqbaron** [comment]
Right, so, crawled out from under my rock and applied the suggestions as I understand them. I put the new functionality in `filter_map` on the logic that it's a special case of that lint's domain, and tried to adopt the suggested methods of detecting interesting functions.

File: clippy_lints/src/methods/filter_map.rs [hunk 11]
```diff
@@ -1,87 +1,167 @@
...
+        let msg = "`filter` for `Some` followed by `unwrap`";
+        let help = "consider using `flatten` instead";
+        let sugg = format!(
+            "{}",
+            reindent_multiline(Cow::Borrowed("flatten()"), true, indent_of(cx, target_span),)
+        );
+        span_lint_and_sugg(
+            cx,
+            OPTION_FILTER_MAP,
+            // expr.span.with_hi(expr.span.hi()),
```
Comment (flip1995) [inline clippy_lints/src/methods/filter_map.rs:79]: NIT:

```suggestion
```

**Review by flip1995 — APPROVED** [review]
Great to hear that you're back! This looks great now. One small clean up left to done, one squash of your commits and this is good to go.

File: clippy_lints/src/methods/filter_map.rs [hunk 12]
```diff
@@ -1,87 +1,167 @@
...
+        _ => false,
+    }
+}
+
+fn is_option_filter_map<'tcx>(
+    cx: &LateContext<'tcx>,
+    filter_arg: &'tcx hir::Expr<'_>,
+    map_arg: &'tcx hir::Expr<'_>,
+) -> bool {
+    is_method(cx, map_arg, sym!(unwrap)) && is_method(cx, filter_arg, sym!(is_some))
```
Comment (flip1995) [inline clippy_lints/src/methods/filter_map.rs:54]: Dogfood says:

```suggestion
    is_method(cx, map_arg, sym::unwrap) && is_method(cx, filter_arg, sym::is_some)
```

(May need to import `rustc_span::sym`

Comment (flip1995) [reply clippy_lints/src/methods/filter_map.rs:54]: Oh yeah, just keep `sym!` here then. I didn't pay much attention to what the dogfood error actually said, sorry :sweat_smile:

Comment (flip1995) [reply clippy_lints/src/methods/filter_map.rs:54]: So keep `sym!` in the `is_some` case and use `sym::` in the `unwrap` case.

Comment (bbqbaron) [reply clippy_lints/src/methods/filter_map.rs:54]: Oh, the dogfood doesn't even suggest `is_some`.

**Comment by flip1995** [comment]
@bors r+

Thanks!
